### PR TITLE
Keypress event queuing

### PR DESCRIPTION
**Pull request by jamiesanson**
This pull request contains code implementing an interrupt driven library for key press event queuing. It implements software debouncing using interval timeouts, and utilizes a callback function implemented in the example main sketch.

### CODE REVIEW

**Comment by PartizanParticle** [comment]
Nice clean code (as expected).

Unsure about the way only 6 key-presses are implemented in hard code, would it be possible to alter this in future so that users can define as many keys as they want?

**Comment by jamiesanson** [comment]
@patricksav Yeah, it's tricky as each individual button needs to have its own ISR assigned to it, else there would be no way to differentiate which button was pressed. I have an idea regarding using lambdas for runtime ISR declaration, so will look into that before merging this.

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
+#include <Arduino.h>
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:1]: Could be wrong but if this is included in header, I don't think it's needed here

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            // Only attach falling edge interrupts for those keys we care about holding down, i.e modifiers
+            if (config.buttonModes[i] == MODIFIER) {
+                attachInterrupt(_config.buttonPins[i], getISR(i), FALLING);
+            }
+        }
+    }
+}
+
+// Setting the callback function used to emit events to
+void ButtonHandler::setListener(EventListener listener) {
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:27]: keep `{` location consistent

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    _listener = listener;
+}
+
+// Emits the next queued event if it exists
+void ButtonHandler::eventComplete() 
+{
+
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:34]: unneeded line break

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// Shifts event array down an indices such that index 1 -> index 0 etc.
+void ButtonHandler::resizeEvents() 
+{
+    for (int i = 0; i < _lastIndex; i++) {
+        _events[i] = _events[i+1];
+    }
+
+    _lastIndex -= 1;
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:56]: is there a convention of when you should use `--`

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    Serial.println(String(_lastIndex));
+}
+
+// Adds event to the queue if there is room in the queue.
+// If the consumer of the library is not processing an event, and the 
+// events queue is empty then simply emit the event, and queue others
+void ButtonHandler::addEvent(ButtonEvent event) {
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:63]: `{`

File: src/libraries/ButtonHandler/ButtonHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        } else {            
+            _events[_lastIndex] = event;            
+            _lastIndex += 1;
+        }
+        interrupts();
+    }
+}
+
+// Region ISR implementation
+void ButtonHandler::onKeyOneEvent()
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.cpp:82]: 🐑  <- lamba

File: src/libraries/ButtonHandler/ButtonHandler.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
+/**
+    This file, along with the accompanying implementation file contain code to handle button press events
+    from simple buttons, where a HI logic level represents a button pressed.
+*/
+
+#include <Arduino.h>
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.h:6]: usually you'd want to include the includes in the declaration guard

File: src/libraries/ButtonHandler/ButtonHandler.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+#include <avr/pgmspace.h>
+
+#ifndef ButtonHandler_H
+#define ButtonHandler_H
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.h:10]: so after this line ^^

Comment (marcelvanworkum) [reply src/libraries/ButtonHandler/ButtonHandler.h:10]: That way the global namespace doesn't get polluted with unneeded imports

File: src/libraries/ButtonHandler/ButtonHandler.h [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+// A sad hack to allow access to member functions from a global scope
+inline void keyOneAction() { _buttonHandler->onKeyOneEvent(); }
+inline void keyTwoAction() { _buttonHandler->onKeyTwoEvent(); }
+inline void keyThreeAction() { _buttonHandler->onKeyThreeEvent(); }
+inline void keyFourAction() { _buttonHandler->onKeyFourEvent(); }
+inline void keyFiveAction() { _buttonHandler->onKeyFiveEvent(); }
+inline void keySixAction() { _buttonHandler->onKeySixEvent(); }
+
+#endif
```
Comment (marcelvanworkum) [inline src/libraries/ButtonHandler/ButtonHandler.h:148]: unix convention to end file with new line

**Comment by marcelvanworkum** [comment]
You should look into #pragma mark comments as well, useful for breaking code up into subsections.

File: src/libraries/ButtonHandler/ButtonHandler.cpp (hunk 1 above)
Comment (jamiesanson) [reply src/libraries/ButtonHandler/ButtonHandler.cpp:1]: Yeah you're right, will remove

File: src/libraries/ButtonHandler/ButtonHandler.cpp (hunk 4 above)
Comment (jamiesanson) [reply src/libraries/ButtonHandler/ButtonHandler.cpp:56]: I think from C paradigms, `--` is apparently not obvious? I remember reading that somewhere. Anyway, I don't think we're gunna ever need to increment a variable by more than 1 at a time, so will change to `++` and `--`

File: src/libraries/ButtonHandler/ButtonHandler.cpp (hunk 6 above)
Comment (jamiesanson) [reply src/libraries/ButtonHandler/ButtonHandler.cpp:82]: `[](memes) -> (memeReturn) {
return memes.get(69);
};`

File: src/libraries/ButtonHandler/ButtonHandler.h (hunk 7 above)
Comment (jamiesanson) [reply src/libraries/ButtonHandler/ButtonHandler.h:6]: Yeah my bad, will shift

File: src/libraries/ButtonHandler/ButtonHandler.h (hunk 9 above)
Comment (jamiesanson) [reply src/libraries/ButtonHandler/ButtonHandler.h:148]: `\n`

**Pull request closed without merging**